[JSC] Add and use StringImpl::isValidLength in toLocale{Lower,Upper}Case

syg opened the pull request:
#### 5771bb5b8ed95fecaf3058d0e2bcf0b4e4c8b7b1
<pre>
[JSC] Add and use StringImpl::isValidLength in toLocale{Lower,Upper}Case
<a href="https://bugs.webkit.org/show_bug.cgi?id=302684">https://bugs.webkit.org/show_bug.cgi?id=302684</a>
<a href="https://rdar.apple.com/157756523">rdar://157756523</a>

Reviewed by Mark Lam.

StringImpl::MaxLength is INT32_MAX, but the real max length is slightly smaller
for 16bit strings to avoid overflowing the unsigned allocation size. This
overflow was fixed a long time ago with manual checked arithmetic.

This difference shows up in toLocale{Lower,Upper}Case, where an 8bit string near
the max length can be inflated to 16bit. Currently we crash instead of throwing
OOM to user code.

This PR does the following:

- Renames StringImpl::maxInternalLength to a public constexpr
  isValidLength&lt;CharType&gt;(size_t) method.
- Throws OOMs in toLocale{Lower,Upper}Case to user code.

No new tests are added as resource exhaustion tests are brittle and slow.

Canonical link: <a href="https://commits.webkit.org/303363@main">https://commits.webkit.org/303363@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d311064240b3ea07c6f55844a3bfdfdd53fa29a4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132157 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/4650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43181 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139672 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/84077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/fc98dda2-d8f2-41a2-858b-b574a6590a5a) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4598 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4411 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/101028 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/84077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f91a8e88-1f60-478b-9ee9-fbdac8238770) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135103 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/118372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/81820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82891 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124220 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/111936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/36496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142319 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130664 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/4319 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37072 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4400 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109580 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27755 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/3275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114648 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/57565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4373 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/33035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163631 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/4205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/67819 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42524 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager (failure)") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/4464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4332 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

syg commented on Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
```diff
@@ -812,11 +812,6 @@ static ALWAYS_INLINE JSString* replaceAllWithCacheUsingRegExpSearchThreeArgument
-    if (totalLength > StringImpl::MaxLength) [[unlikely]] {
-        throwOutOfMemoryError(globalObject, scope);
-        return nullptr;
-    }
```
AFAICT this check is superfluous because `totalLength` gets checked in the `StringImpl::tryCreateUninitialized` call below.

syg commented on Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
```diff
@@ -957,11 +952,6 @@ static ALWAYS_INLINE JSString* replaceAllWithCacheUsingRegExpSearch(VM& vm, JSGl
-        if (totalLength > StringImpl::MaxLength) [[unlikely]] {
-            throwOutOfMemoryError(globalObject, scope);
-            return nullptr;
-        }
```
AFAICT this check is superfluous because `totalLength` gets checked in the `StringImpl::tryCreateUninitialized` call below.

syg commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -416,8 +417,6 @@ Ref<StringImpl> StringImpl::convertToLowercaseWithoutLocale()
     }
 
-    if (m_length > MaxLength)
-        CRASH();
```
This check is superfluous because `m_length` is checked in `createUninitializedInternalNonEmpty` below.

MenloDorian reviewed (commented):
typo in commit msg: "This different shows up " => "This difference shows up "

MenloDorian commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -1399,15 +1386,14 @@ Ref<StringImpl> StringImpl::replace(StringView pattern, StringView replacement)
 
+    Checked<size_t> newSize = m_length;
+    newSize -= matchSize;
+    newSize += replaceSize;
```
This will still crash on overflow (as the old code did).  That is your intention correct?

syg replied:
Yes. Other than the toLocaleCase change, everything else is designed to be semantics preserving wrt crashing.

MenloDorian reviewed (approved):
r=me.

syg committed:
[JSC] Add and use StringImpl::isValidLength in toLocale{Lower,Upper}Case
https://bugs.webkit.org/show_bug.cgi?id=302684
rdar://157756523

Reviewed by Mark Lam.

StringImpl::MaxLength is INT32_MAX, but the real max length is slightly smaller
for 16bit strings to avoid overflowing the unsigned allocation size. This
overflow was fixed a long time ago with manual checked arithmetic.

This difference shows up in toLocale{Lower,Upper}Case, where an 8bit string near
the max length can be inflated to 16bit. Currently we crash instead of throwing
OOM to user code.

This PR does the following:

- Renames StringImpl::maxInternalLength to a public constexpr
  isValidLength<CharType>(size_t) method.
- Throws OOMs in toLocale{Lower,Upper}Case to user code.

No new tests are added as resource exhaustion tests are brittle and slow.

Canonical link: https://commits.webkit.org/303363@main
